### PR TITLE
FilterPathBasedFilter support match fieldname with dot

### PR DESCRIPTION
**Pull request by mushao999**
Current `FilterPathBasedFilter` does not support match fieldName with dot.
This PR merges the changes of #83148 and #83152 together to add the ability of `matching fieldName with dot` to the `FilterPathBasedFilter`.

### CODE REVIEW

**Comment by nik9000** [comment]
@elasticmachine, test this please

**Review by nik9000 — CHANGES_REQUESTED** [review]
I think you've added two bits of my other PR that you should remove. I commented on the files. *I* think they are the right thing to change, but I want to get them reviewed by other folks. But I can review the xcontent change. I think it fine - though I'd prefer a non-recursive solution. I added a couple other test cases, one that blows up in a recursive solution.

I see now what you are doing with prefix matching - it's for if the pattern *fully* matches. That makes sense to me. I added some more test cases to the abstract filtering test case thingy for that too.

File: server/src/main/java/org/elasticsearch/cluster/routing/IndexRouting.java [hunk 1]
```diff
@@ -8,6 +8,8 @@
 
 package org.elasticsearch.cluster.routing;
 
+import org.apache.lucene.util.BytesRef;
+import org.apache.lucene.util.StringHelper;
```
[the remark this answers is not in the document]
Comment (mushao999) [reply server/src/main/java/org/elasticsearch/cluster/routing/IndexRouting.java]: upated

**Comment by mushao999** [comment]
@nik9000 I've rollback the changes you comment and add some tests for double wildcard in `AbstractXContentFilteringTestCase`.

**Comment by nik9000** [comment]
@elasticmachine, test this please

**Comment by nik9000** [comment]
Thanks for your changes so far! I think one of the tests is still going to fail because the solution is recursive, but otherwise I'm happy with it. Could you try running `JsonXContentFilteringTests` and see what that does? I *thought* the test I pushed would fail. At least, it did on my laptop yesterday.

**Comment by nik9000** [comment]
I talked with @mushao999 about the recursion here. He mentioned that the test that I added actually stack overflows on other code! Ooops. But he's going to zap his recursion to make me happy. I'm thankful for that.

**Comment by nik9000** [comment]
@elasticmachine, test this please

**Review by nik9000 — APPROVED** [review]
For now we're going to let the recursion live. We both want to zap it, but don't have any good ideas at the moment.

**Pull request merged**